ee/codegen/src/generators/graph-attribute.ts

```diff
@@ -110,6 +149,157 @@ export class GraphAttribute extends AstNode {
     return graphMutableAst;
   }
 
+  private groupEdgesByConnectedComponents(
+    edges: WorkflowEdge[]
+  ): WorkflowEdge[][] {
+    const components: WorkflowEdge[][] = [];
+    const edgeVisited = new Set<string>();
+    const adjacencyList = new Map<string, WorkflowEdge[]>();
+
+    edges.forEach((edge) => {
+      adjacencyList.set(edge.sourceNodeId, [
+        ...(adjacencyList.get(edge.sourceNodeId) || []),
+        edge,
+      ]);
+      adjacencyList.set(edge.targetNodeId, [
+        ...(adjacencyList.get(edge.targetNodeId) || []),
+        edge,
+      ]);
+    });
+
+    for (const edge of edges) {
+      if (edgeVisited.has(edge.id)) continue;
+
+      const component: WorkflowEdge[] = [];
+      const queue: WorkflowEdge[] = [edge];
+
+      while (queue.length > 0) {
+        const currentEdge = queue.shift();
+        if (!currentEdge || edgeVisited.has(currentEdge.id)) continue;
+
+        edgeVisited.add(currentEdge.id);
+        component.push(currentEdge);
+
+        adjacencyList.get(currentEdge.sourceNodeId)?.forEach((nextEdge) => {
+          if (!edgeVisited.has(nextEdge.id)) queue.push(nextEdge);
+        });
+
+        adjacencyList.get(currentEdge.targetNodeId)?.forEach((nextEdge) => {
+          if (!edgeVisited.has(nextEdge.id)) queue.push(nextEdge);
+        });
+      }
+
+      components.push(component);
+    }
+
+    return components;
+  }
+
+  private findRootNodes(componentEdges: WorkflowEdge[]): string[] {
+    const incomingEdges = new Map<string, number>();
+
+    // Count incoming edges
+    componentEdges.forEach((edge) => {
+      incomingEdges.set(
+        edge.targetNodeId,
+        (incomingEdges.get(edge.targetNodeId) || 0) + 1
+      );
+      if (!incomingEdges.has(edge.sourceNodeId)) {
+        incomingEdges.set(edge.sourceNodeId, 0);
+      }
+    });
+
+    // Return nodes with zero incoming edges
+    return Array.from(incomingEdges.entries())
```

Comment:
[nit] Do we need to do this over `incomingEdges`? Or can we just iterate over nodes left that have no incoming edge?